### PR TITLE
Probe type bias correction (and other preprocessing) for 27K methylation data

### PR DESCRIPTION
**Pull request by jjc2718**
Methylation arrays contain 2 different types of CpG site probes (called type I and type II) that work slightly differently: see, e.g. [the intro to this paper](https://www.sciencedirect.com/science/article/pii/S1046202314002692) for more information. In analyses of methylation data, it's somewhat common to correct for differing probe intensities within each sample. This ensures that there's not a bias toward one type of probe or the other, which can be particularly problematic in differential methylation analyses.

Previously, we weren't doing any correction for probe type bias, but we wanted to try it to get an idea of whether or not it matters. We applied bias correction to the 27K data to start with since it's a much smaller dataset. We chose the [BMIQ method](https://academic.oup.com/bioinformatics/article/29/2/189/204142) to do the correction, since it has an up-to-date R implementation and is commonly used in other similar studies.

Our correction does seem to work (type II probes have similar beta distributions to type I probes within samples):

image

But when we use the resulting data to do mutation prediction, it doesn't seem to make much of a difference:

image

So going forward, we'll likely just use the uncorrected data.

### CODE REVIEW

**Review by ben-heil — APPROVED** [review]
It seems like it's worth eyeballing the difference in distributions between the corrected and uncorrected methylation data's distributions. I feel like it's easier to argue "correction doesn't change anything for this dataset" than "because this classifier had the same performance before/after correction, correction doesn't do anything".

It's also possible that you've already made that plot and it just didn't make the notebook. In that case, disregard this comment

File: 00_download_data/3_methylation_beta.ipynb [hunk 1]
```diff
@@ -0,0 +1,1018 @@
...
+    "\n",
+    "pca_dir = os.path.join(cfg.data_dir, 'me_compressed')\n",
+    "os.makedirs(pca_dir, exist_ok=True)\n",
+    "\n",
+    "tcga_processed_df = (tcga_norm_df\n",
+    "  .drop(columns=['probe_type', 'chromosome'])\n",
+    "  .transpose()\n",
+    ")\n",
+    "\n",
+    "if SAVE_RESULTS:\n",
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply 00_download_data/3_methylation_beta.ipynb:835]: It takes a long time to generate the PCA results (especially for 5000 PCs), so I was only generating them if I wanted to save them. 

I guess if `SAVE_RESULTS=False` I could load the saved results instead and plot them - I'll try doing that.

File: 02_classify_stratified/nbconverted/beta_correction.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+            zorder=0)
+    ax.set_xlabel('AUPR({}) - AUPR({})'.format(train_2, train_1), size=14)
+    ax.set_ylabel(r'$-\log_{10}($adjusted $p$-value$)$', size=14)
+    ax.set_xlim(xlim)
+    ax.set_ylim(ylim)
+    ax.legend(title=r'Reject $H_0$', loc='upper left')
+    ax.set_title(r'Mutation prediction, {} vs. {}'.format(train_2, train_1), size=14)
+
+    text_labels = label_points(compare_results_df['delta_mean'],
+                               compare_results_df['nlog10_p'],
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply 02_classify_stratified/nbconverted/beta_correction.py:214]: Yeah, that's exactly it. The x-axis is the difference in means (across CV folds) between the corrected and non-corrected data, and the y-axis is the p-value from a t-test comparing the CV results distributions. Each point is a gene.

So the fact that all the results on the far right plot are clustered near the origin (small difference in means, very small p-values) means the predictive performance is essentially the same before and after correction for all the genes in our gene set.

I'll add this to the script too.

**Comment by jjc2718** [comment]
> It seems like it's worth eyeballing the difference in distributions between the corrected and uncorrected methylation data's distributions. I feel like it's easier to argue "correction doesn't change anything for this dataset" than "because this classifier had the same performance before/after correction, correction doesn't do anything".
> 
> It's also possible that you've already made that plot and it just didn't make the notebook. In that case, disregard this comment

I did make the plot, but I haven't thought hard about how to interpret it until now. You can see the difference between the corrected and uncorrected distributions in [the first slide here](https://docs.google.com/presentation/d/1vz8MOxObA7ESgiGhNtP2C3mpUotT8PyTkwRyXKdv8RM/edit?usp=sharing), and I'll also add this to the notebook.

I don't quite think we want to argue that correction doesn't change anything for this dataset. We were hoping that would be the case (since we're using TCGA level 3 data which does some types of normalization/batch correction already), but it does look like this bias correction method changes the distributions noticeably.

The goal of the method we're using is not to make the type I and type II distributions exactly the same, but to make the type I and type II probes within each sample fit similar distributions (preserving rank, etc) while maintaining biologically relevant variation. You can particularly see this in the second column of the slide I linked above - the means for the modes of each distribution are the same after correction, and the individual intensities per probe are a bit closer together than they were before. In the first column the average intensities (averaged over all probes) aren't the same, but this likely reflects actual biological variation between type I and type II probes. In the [original BMIQ paper](https://academic.oup.com/bioinformatics/article/29/2/189/204142) they discuss similar observations, but this is a really complicated problem (that I don't fully understand TBH) and we're largely okay with just trusting whatever they're doing.

So, with that said, I think I would rephrase what you said here:

> "because this classifier had the same performance before/after correction, correction doesn't do anything"

to something more like this:

> "this classifier had the same performance before/after correction, so correction likely doesn't matter for this problem"

and we think the results in `beta_correction.ipynb` largely supports that. If we saw more of a difference we'd probably dive deeper into individual correction methods and how they work, but our goal was just to check if it seemed to matter in a relatively limited case, and we're okay with saying "probably not" and moving on based on what we see here.

File: 00_download_data/3_methylation_beta.ipynb [hunk 3]
```diff
@@ -0,0 +1,1018 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Normalize type I vs. type II beta values\n",
+    "\n",
+    "This happens on the sample level.\n",
+    "\n",
+    "At this point, you should run the `run_bmiq.R` script in the `wateRmelon` conda environment. This should generate the `methylation_27k_bmiq_normalized.tsv` output that we load below."
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply 00_download_data/3_methylation_beta.ipynb:648]: See comment below and linked slides (I'll also add it to this notebook)
